modules/sitemaps/sitemap-buffer.php

```diff
@@ -0,0 +1,287 @@
+	 *
+	 * @access private
+	 * @since 4.7.0
+	 * @var int $byte_capacity The byte capacity.
+	 */
+	private $byte_capacity;
+
+	/**
+	 * Footer text of the buffer; stored here so it can be appended when the buffer is full.
+	 *
+	 * @access private
+	 * @since 4.7.0
+	 * @var string $footer_text The footer text.
+	 */
+	private $footer_text;
+
+	/**
+	 * The buffer contents.
+	 *
+	 * @access private
+	 * @since 4.7.0
+	 * @var string The buffer contents.
+	 */
+	private $buffer;
+
+	/**
+	 * Flag which detects when the buffer is full.
+	 *
+	 * @access private
+	 * @since 4.7.0
+	 * @var bool $is_full_flag The flag value. This flag is set to false on construction and only flipped to true if we've tried to add something and failed.
+	 */
+	private $is_full_flag;
+
+	/**
+	 * Flag which detects when the buffer is empty.
+	 *
+	 * @access private
+	 * @since 4.7.0
+	 * @var bool $is_empty_flag The flag value. This flag is set to true on construction and only flipped to false if we've tried to add something and succeeded.
+	 */
+	private $is_empty_flag;
+
+	/**
+	 * The most recent timestamp seen by the buffer.
+	 *
+	 * @access private
+	 * @since 4.7.0
+	 * @var string $timestamp Must be in 'YYYY-MM-DD hh:mm:ss' format.
+	 */
+	private $timestamp;
+
+	/**
+	 * Construct a new Jetpack_Sitemap_Buffer.
+	 *
+	 * @since 4.7.0
+	 *
+	 * @param int    $item_limit The maximum size of the buffer in items. Default is 50000.
+	 * @param int    $byte_limit The maximum size of the buffer in bytes. Default is 10485760 (10MB).
+	 * @param string $header The string to prepend to the entire buffer.
+	 * @param string $footer The string to append to the entire buffer.
+	 * @param string $time The initial timestamp of the buffer. Must be in 'YYYY-MM-DD hh:mm:ss' format.
+	 */
+	public function __construct(
+		$item_limit,
+		$byte_limit,
+		$header = '',
+		$footer = '',
+		$time
+	) {
+		$this->item_capacity = max( 1, intval( $item_limit ) );
+
+		mbstring_binary_safe_encoding(); // So we can safely use mb_strlen().
+		$this->byte_capacity = $byte_limit - mb_strlen( $header ) - mb_strlen( $footer );
+		reset_mbstring_encoding();
+
+		$this->footer_text = $footer;
+		$this->buffer = $header;
+		$this->is_full_flag = false;
+		$this->is_empty_flag = true;
+		$this->timestamp = $time;
+		return;
+	}
+
+	/**
+	 * Append an item to the buffer, if there is room for it.
+	 * If not, we set is_full_flag to true. If $item is null,
+	 * don't do anything and report success.
+	 *
+	 * @since 4.7.0
+	 *
+	 * @param string $item The item to be added.
+	 *
+	 * @return bool True if the append succeeded, False if not.
+	 */
+	public function try_to_add_item( $item ) {
+		if ( is_null( $item ) ) {
+			return true;
+		} else {
+			mbstring_binary_safe_encoding(); // So we can safely use mb_strlen().
+
+			if ( 0 >= $this->item_capacity || 0 > $this->byte_capacity - mb_strlen( $item ) ) {
+				$this->is_full_flag = true;
+				reset_mbstring_encoding();
```

Comment:
There's a similar issue [here](https://github.com/nbloomf/jetpack/blob/853515f6eb8a89c117515ac1297da7fa16981711/modules/sitemaps/sitemap-builder.php#L122), where we lock/unlock the sitemap state; the state is unlocked once but locked in several places. I'd like to clean that up too.